// asbuild/dist/commands/fmt.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.initConfig = exports.FmtCmd = exports.fmtCmdBuilder = void 0;
var path = __importStar(require("path"));
var eslintConfig_1 = require("./init/files/eslintConfig");
var interfaces_1 = require("./init/interfaces");
var chalk_1 = __importDefault(require("chalk"));
var packageJson_1 = require("./init/files/packageJson");
var eslint_1 = require("eslint");
var utils_1 = require("../utils");
// TODO:
// - Support more eslint options, like '--max-warnings`,
//   (Much of these can be ported from eslint's `cli.js`)
exports.fmtCmdBuilder = function (y) {
    return y
        .positional("paths", {
        description: "Paths to format",
        default: ["."],
    })
        .option("init", {
        description: "Generates recommended eslint config for AS Projects",
        boolean: true,
        group: "Initialisation:",
    })
        .option("lint", {
        alias: ["dry-run"],
        boolean: true,
        default: false,
        description: "Tries to fix problems without saving the changes to the file system",
        group: "Miscellaneous",
    });
};
exports.FmtCmd = {
    command: "fmt [paths..]",
    describe: "This utility formats current module using eslint.",
    aliases: ["format", "lint"],
    builder: function (y) {
        return exports.fmtCmdBuilder(y).onFinishCommand(function (code) { return process.exit(code); });
    },
    handler: function (args) { return __awaiter(void 0, void 0, void 0, function () {
        var retCode, files, engine, results, formatter, resultText, error_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (args.init) {
                        return [2 /*return*/, initConfig(process.cwd())];
                    }
                    retCode = 0;
                    files = args.paths;
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 6, , 7]);
                    engine = new eslint_1.ESLint({
                        extensions: ["ts"],
                        fix: true,
                    });
                    return [4 /*yield*/, engine.lintFiles(files)];
                case 2:
                    results = _a.sent();
                    if (!!args.dryRun) return [3 /*break*/, 4];
                    // fix files in place
                    return [4 /*yield*/, eslint_1.ESLint.outputFixes(results)];
                case 3:
                    // fix files in place
                    _a.sent();
                    _a.label = 4;
                case 4: return [4 /*yield*/, engine.loadFormatter("stylish")];
                case 5:
                    formatter = _a.sent();
                    resultText = formatter.format(results);
                    utils_1.log(resultText);
                    utils_1.log(chalk_1.default(templateObject_1 || (templateObject_1 = __makeTemplateObject(["{bold.green Done!}"], ["{bold.green Done!}"]))));
                    return [3 /*break*/, 7];
                case 6:
                    error_1 = _a.sent();
                    utils_1.log(chalk_1.default(templateObject_2 || (templateObject_2 = __makeTemplateObject(["{bold.bgRedBright ERROR:} Unexpected Error while running ESlint on given files."], ["{bold.bgRedBright ERROR:} Unexpected Error while running ESlint on given files."]))), true);
                    utils_1.log(error_1, true);
                    retCode = 1;
                    return [3 /*break*/, 7];
                case 7: return [2 /*return*/, retCode];
            }
        });
    }); },
};
function initConfig(baseDir) {
    // write the config file
    var dir = path.resolve(baseDir);
    var eslintFile = new eslintConfig_1.EslintConfigFile();
    var res = interfaces_1.InitResult.CREATED;
    utils_1.log(chalk_1.default(templateObject_3 || (templateObject_3 = __makeTemplateObject(["Writing {cyan ", "} ..."], ["Writing {cyan ", "} ..."])), eslintFile.getRelativePath(dir)));
    switch (eslintFile.write(dir)) {
        case interfaces_1.InitResult.EXISTS:
            res = interfaces_1.InitResult.EXISTS;
            utils_1.log(chalk_1.default(templateObject_4 || (templateObject_4 = __makeTemplateObject(["File {bold.cyan ", "} already exists."], ["File {bold.cyan ",
                "} already exists."])), eslintFile.getRelativePath(dir)));
            break;
        case interfaces_1.InitResult.CREATED:
            utils_1.log([
                chalk_1.default(templateObject_5 || (templateObject_5 = __makeTemplateObject(["{bold.green Created:} ", ""], ["{bold.green Created:} ", ""])), eslintFile.path),
                "",
                chalk_1.default(templateObject_6 || (templateObject_6 = __makeTemplateObject(["{bold.green Done!}"], ["{bold.green Done!}"]))),
                chalk_1.default(templateObject_7 || (templateObject_7 = __makeTemplateObject([""], [""]))),
                chalk_1.default(templateObject_8 || (templateObject_8 = __makeTemplateObject(["Don't forget to install 'eslint' and it's plugins before you start:"], ["Don't forget to install 'eslint' and it's plugins before you start:"]))),
                "",
                chalk_1.default(templateObject_9 || (templateObject_9 = __makeTemplateObject(["  ", " eslint"], ["  ", " eslint"])), packageJson_1.getPmCommands().pkgInstall),
                chalk_1.default(templateObject_10 || (templateObject_10 = __makeTemplateObject(["  ", " @typescript-eslint/parser"], ["  ", " @typescript-eslint/parser"])), packageJson_1.getPmCommands().pkgInstall),
                chalk_1.default(templateObject_11 || (templateObject_11 = __makeTemplateObject(["  ", " @typescript-eslint/eslint-plugin"], ["  ",
                    " @typescript-eslint/eslint-plugin"])), packageJson_1.getPmCommands().pkgInstall),
                "",
                chalk_1.default(templateObject_12 || (templateObject_12 = __makeTemplateObject(["Have a nice day !"], ["Have a nice day !"]))),
            ].join("\n"));
            break;
        default:
            break;
    }
    return res;
}
exports.initConfig = initConfig;
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6, templateObject_7, templateObject_8, templateObject_9, templateObject_10, templateObject_11, templateObject_12;
