// asbuild/dist/utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.mockModule = exports.log = exports.askYesNo = exports.ensureDirExists = exports.getGlobalCliCallback = exports.setGlobalCliCallback = exports.getGlobalAscOptions = exports.setGlobalAscOptions = void 0;
var path = __importStar(require("path"));
var fs = __importStar(require("fs"));
var readline = __importStar(require("readline"));
var DEFAULT_ASC_OPTIONS = {};
var DEFAULT_CLI_CALLBACK = function (err) {
    if (err) {
        throw err;
    }
    return 1;
};
// @ts-ignore
var ASC_OPTIONS = DEFAULT_ASC_OPTIONS;
// @ts-ignore
var CLI_CALLBACK = DEFAULT_CLI_CALLBACK;
/**
 * @note Only for use by testing mechanism
 */
function setGlobalAscOptions(ascOptions) {
    if (ascOptions === void 0) { ascOptions = DEFAULT_ASC_OPTIONS; }
    ASC_OPTIONS = ascOptions;
}
exports.setGlobalAscOptions = setGlobalAscOptions;
function getGlobalAscOptions() {
    return ASC_OPTIONS;
}
exports.getGlobalAscOptions = getGlobalAscOptions;
/**
 * @note Only for use by testing mechanism
 */
function setGlobalCliCallback(callback) {
    if (callback === void 0) { callback = DEFAULT_CLI_CALLBACK; }
    CLI_CALLBACK = callback;
}
exports.setGlobalCliCallback = setGlobalCliCallback;
function getGlobalCliCallback() {
    return CLI_CALLBACK;
}
exports.getGlobalCliCallback = getGlobalCliCallback;
function ensureDirExists(filePath) {
    var dirname = path.dirname(filePath);
    if (fs.existsSync(dirname)) {
        return;
    }
    ensureDirExists(dirname);
    fs.mkdirSync(dirname);
}
exports.ensureDirExists = ensureDirExists;
function askYesNo(ques) {
    return __awaiter(this, void 0, void 0, function () {
        var rl, res;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    rl = readline.createInterface({
                        input: process.stdin,
                        output: process.stdout,
                        terminal: false,
                    });
                    return [4 /*yield*/, new Promise(function (resolve) {
                            return rl.question(ques + " [Y/n]: ", resolve);
                        })];
                case 1:
                    res = _a.sent();
                    rl.close();
                    return [2 /*return*/, /^$|^y?$/.test(res.toLowerCase().trim())];
            }
        });
    });
}
exports.askYesNo = askYesNo;
function log(message, error) {
    if (error === void 0) { error = false; }
    error ? console.error(message) : console.info(message);
}
exports.log = log;
function mockModule(moduleToMock, defaultMockValuesForMock) {
    return function (sandbox, returnOverrides) {
        var returns = returnOverrides || {};
        var functions = Object.keys(moduleToMock);
        functions.forEach(function (f) {
            var override = returns[f] || defaultMockValuesForMock[f];
            if (override) {
                sandbox.stub(moduleToMock, f).callsFake(override);
            }
        });
    };
}
exports.mockModule = mockModule;
