// asbuild/dist/commands/init/files/asconfigJson.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.AsconfigJsonFile = void 0;
var interfaces_1 = require("../interfaces");
var AsconfigJsonFile = /** @class */ (function (_super) {
    __extends(AsconfigJsonFile, _super);
    function AsconfigJsonFile() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.path = "asconfig.json";
        _this.description = "Configuration file defining both a 'debug' and a 'release' target.";
        _this.configObj = {
            targets: {
                debug: {
                    binaryFile: "build/untouched.wasm",
                    textFile: "build/untouched.wat",
                    sourceMap: true,
                    debug: true,
                },
                release: {
                    binaryFile: "build/optimized.wasm",
                    textFile: "build/optimized.wat",
                    sourceMap: true,
                    optimizeLevel: 3,
                    shrinkLevel: 1,
                    converge: false,
                    noAssert: false,
                },
            },
            options: {},
        };
        _this.updateOldContent = null;
        return _this;
    }
    AsconfigJsonFile.prototype.getContent = function () {
        return JSON.stringify(this.configObj, null, 2);
    };
    return AsconfigJsonFile;
}(interfaces_1.InitFile));
exports.AsconfigJsonFile = AsconfigJsonFile;
