// asbuild/dist/commands/init/cmd.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.writeFiles = exports.InitCmd = exports.initCmdBuilder = void 0;
var files_1 = require("./files");
var utils_1 = require("../../utils");
var interfaces_1 = require("./interfaces");
var path = __importStar(require("path"));
var chalk_1 = __importDefault(require("chalk"));
var packageJson_1 = require("./files/packageJson");
var doneMsg = chalk_1.default(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\nDon't forget to install dependencies before you start:\n\n  ", "\n\nTo edit the entry file, open {cyan 'assembly/index.ts'} in your editor of choice.\nCreate as many additional files as necessary and use them as imports.\n\nTo build the entry file to WebAssembly when you are ready, run:\n\n  ", " build\n\nRunning the command above creates the following binaries incl. their respective\ntext format representations and source maps:\n\n  {cyan ./build/untouched.wasm}\n  {cyan ./build/untouched.wasm.map}\n  {cyan ./build/untouched.wat}\n\n  ^ The untouched WebAssembly module as generated by the compiler.\n    This one matches your sources exactly, without any optimizations.\n\n  {cyan ./build/optimized.wasm}\n  {cyan ./build/optimized.wasm.map}\n  {cyan ./build/optimized.wat}\n\n  ^ The optimized WebAssembly module using default optimization settings.\n    You can change the optimization settings in {cyan 'package.json'}.\n\nTo run the tests, do:\n\n  ", "\n\nThe AssemblyScript documentation covers all the details:\n\n  https://docs.assemblyscript.org\n\nHave a nice day!\n"], ["\nDon't forget to install dependencies before you start:\n\n  ", "\n\nTo edit the entry file, open {cyan 'assembly/index.ts'} in your editor of choice.\nCreate as many additional files as necessary and use them as imports.\n\nTo build the entry file to WebAssembly when you are ready, run:\n\n  ", " build\n\nRunning the command above creates the following binaries incl. their respective\ntext format representations and source maps:\n\n  {cyan ./build/untouched.wasm}\n  {cyan ./build/untouched.wasm.map}\n  {cyan ./build/untouched.wat}\n\n  ^ The untouched WebAssembly module as generated by the compiler.\n    This one matches your sources exactly, without any optimizations.\n\n  {cyan ./build/optimized.wasm}\n  {cyan ./build/optimized.wasm.map}\n  {cyan ./build/optimized.wat}\n\n  ^ The optimized WebAssembly module using default optimization settings.\n    You can change the optimization settings in {cyan 'package.json'}.\n\nTo run the tests, do:\n\n  ", "\n\nThe AssemblyScript documentation covers all the details:\n\n  https://docs.assemblyscript.org\n\nHave a nice day!\n"])), packageJson_1.getPmCommands().install, packageJson_1.getPmCommands().run, packageJson_1.getPmCommands().test);
function initMsg(baseDir) {
    var msg = [
        chalk_1.default(templateObject_2 || (templateObject_2 = __makeTemplateObject(["This command will make sure that the following files exist in the project"], ["This command will make sure that the following files exist in the project"]))),
        chalk_1.default(templateObject_3 || (templateObject_3 = __makeTemplateObject(["directory {bold \"", "\":}"], ["directory {bold \"", "\":}"])), baseDir),
        "",
    ];
    for (var _i = 0, initFiles_1 = files_1.initFiles; _i < initFiles_1.length; _i++) {
        var file = initFiles_1[_i];
        msg.push(chalk_1.default(templateObject_4 || (templateObject_4 = __makeTemplateObject(["  {cyan ./", "}"], ["  {cyan ./", "}"])), file.path));
        msg.push(chalk_1.default(templateObject_5 || (templateObject_5 = __makeTemplateObject(["  ", ""], ["  ", ""])), file.description));
        msg.push("");
    }
    msg.push("The command will try to update existing files to match the correct settings");
    msg.push("");
    return msg.join("\n");
}
exports.initCmdBuilder = function (y) {
    return y
        .positional("baseDir", {
        type: "string",
        default: ".",
        description: "Create a sample AS project in this directory",
    })
        .option("yes", {
        boolean: true,
        default: false,
        description: "Skip the interactive prompt",
    });
};
exports.InitCmd = {
    command: "init [baseDir]",
    describe: "Create a new AS package in an given directory",
    builder: function (y) {
        return exports.initCmdBuilder(y).onFinishCommand(function (code) { return process.exit(code); });
    },
    handler: function (args) { return __awaiter(void 0, void 0, void 0, function () {
        var retCode, baseDir, _a;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    retCode = 0;
                    baseDir = path.resolve(args.baseDir);
                    // print initMsg
                    utils_1.log(initMsg(baseDir));
                    _a = !args.yes;
                    if (!_a) return [3 /*break*/, 2];
                    return [4 /*yield*/, utils_1.askYesNo(chalk_1.default(templateObject_6 || (templateObject_6 = __makeTemplateObject(["{bold Do you to continue?}"], ["{bold Do you to continue?}"]))))];
                case 1:
                    _a = !(_b.sent());
                    _b.label = 2;
                case 2:
                    // ask confirmation, return if 'No'
                    if (_a)
                        return [2 /*return*/, 2];
                    try {
                        writeFiles(baseDir);
                        utils_1.log(doneMsg);
                    }
                    catch (err) {
                        // https://github.com/yargs/yargs/issues/1069#issuecomment-468451450
                        // If error is thrown in async handler, parser is run twice which
                        // cause the help commands to print along with error stack
                        // So to prevent that we catch the error and simply print it.
                        utils_1.log(chalk_1.default(templateObject_7 || (templateObject_7 = __makeTemplateObject(["{bold.bgRedBright ERROR:} Unexpected Error while trying to write init files."], ["{bold.bgRedBright ERROR:} Unexpected Error while trying to write init files."]))), true);
                        utils_1.log(err, true);
                        retCode = 1;
                    }
                    return [2 /*return*/, retCode];
            }
        });
    }); },
};
function writeFiles(baseDir) {
    utils_1.log("");
    for (var _i = 0, initFiles_2 = files_1.initFiles; _i < initFiles_2.length; _i++) {
        var file = initFiles_2[_i];
        utils_1.log(chalk_1.default(templateObject_8 || (templateObject_8 = __makeTemplateObject(["- Making sure that '", "' is set up..."], ["- Making sure that '", "' is set up..."])), file.path));
        var relativePath = file.getRelativePath(baseDir);
        switch (file.write(baseDir)) {
            case interfaces_1.InitResult.CREATED:
                utils_1.log(chalk_1.default(templateObject_9 || (templateObject_9 = __makeTemplateObject(["  {green Created:} ", "\n"], ["  {green Created:} ", "\\n"])), relativePath));
                break;
            case interfaces_1.InitResult.EXISTS:
                utils_1.log(chalk_1.default(templateObject_10 || (templateObject_10 = __makeTemplateObject(["  {yellow Exists:} ", "\n"], ["  {yellow Exists:} ", "\\n"])), relativePath));
                break;
            case interfaces_1.InitResult.UPDATED:
                utils_1.log(chalk_1.default(templateObject_11 || (templateObject_11 = __makeTemplateObject(["  {greenBright Updated:} ", "\n"], ["  {greenBright Updated:} ", "\\n"])), relativePath));
                break;
            default:
                break;
        }
    }
    utils_1.log(chalk_1.default(templateObject_12 || (templateObject_12 = __makeTemplateObject(["{bold.green Done!}"], ["{bold.green Done!}"]))));
}
exports.writeFiles = writeFiles;
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6, templateObject_7, templateObject_8, templateObject_9, templateObject_10, templateObject_11, templateObject_12;
