// asbuild/dist/commands/init/interfaces.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InitFile = exports.InitResult = void 0;
var path = __importStar(require("path"));
var fs = __importStar(require("fs"));
var utils_1 = require("../../utils");
var InitResult;
(function (InitResult) {
    InitResult[InitResult["CREATED"] = 0] = "CREATED";
    InitResult[InitResult["UPDATED"] = 1] = "UPDATED";
    InitResult[InitResult["EXISTS"] = 2] = "EXISTS";
})(InitResult = exports.InitResult || (exports.InitResult = {}));
var InitFile = /** @class */ (function () {
    function InitFile() {
    }
    InitFile.prototype.getRelativePath = function (baseDir) {
        return path.join(baseDir, this.path);
    };
    /**
     * Write the InitFile to given baseDir
     * @param baseDir Base directory where file will created in relative to
     */
    InitFile.prototype.write = function (baseDir) {
        var filePath = this.getRelativePath(baseDir);
        // create the required dirs if not exists
        utils_1.ensureDirExists(filePath);
        // check if file already exists
        var fileExists = fs.existsSync(filePath);
        // check whether file can be updated or not
        var shouldUpdate = this.updateOldContent && fileExists;
        if (fileExists && !shouldUpdate) {
            // if file exists and cannot update return
            return InitResult.EXISTS;
        }
        else {
            var newContent = shouldUpdate && this.updateOldContent
                ? this.updateOldContent(fs.readFileSync(filePath, { encoding: "utf8" }))
                : this.getContent();
            fs.writeFileSync(filePath, newContent);
        }
        return fileExists && shouldUpdate ? InitResult.UPDATED : InitResult.CREATED;
    };
    return InitFile;
}());
exports.InitFile = InitFile;
